okhttp/src/main/java/okhttp3/internal/ws/WebSocketExtensions.kt

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.ws
+
+import java.io.IOException
+import okhttp3.Response
+import okhttp3.internal.delimiterOffset
+import okhttp3.internal.trimSubstring
+
+/**
+ * Models the contents of a `Sec-WebSocket-Extensions` response header. OkHttp honors one extension
+ * `permessage-deflate` and two of its parameters, `client_no_context_takeover` and
+ * `server_no_context_takeover`.
+ *
+ * Typically this will look like one of the following:
+ *
+ * ```
+ * Sec-WebSocket-Extensions: permessage-deflate
+ * Sec-WebSocket-Extensions: permessage-deflate; client_no_context_takeover
+ * Sec-WebSocket-Extensions: permessage-deflate; server_no_context_takeover
+ * Sec-WebSocket-Extensions: permessage-deflate; server_no_context_takeover; client_no_context_takeover
+ * ```
+ *
+ * If any other extension or parameter is specified, then [unknownValues] will be true. Such
+ * responses should be refused as their web socket extensions will not be understood.
+ *
+ * Note that the `client_max_window_bits` and `server_max_window_bits` parameters are unknown
+ * because we do not expect servers to return them. (OkHttp is unable to handle any number of bits
+ * other than 15 (32 KiB) due to a lack of configurability in [java.util.zip.Deflater].)
+ *
+ * See [RFC 7692, 7.1][rfc_7692] for details on negotiation process.
+ *
+ * [rfc_7692]: https://tools.ietf.org/html/rfc7692#section-7.1
+ */
+data class WebSocketExtensions(
```

Comment:
Indeed, with PR separated into several smaller ones, it's much easier to reason about them. 👍